File: Problems/040-combination-sum-ii/combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates, target: int):
        def backtrack(remain, start, end):
            if remain == 0:
                if temp:
                    res.append(temp + [])
                return
            
            for i in range(start, end):
                if i > start and candidates[i] == candidates[i-1]:
                    continue
                remain -= candidates[i]
                if remain < 0:
                    break
                temp.append(candidates[i])
                backtrack(remain, i+1, end)
                remain += temp.pop()
        
        res, temp = [], []
        candidates.sort()
        backtrack(target, 0, len(candidates))
        
        return res

print(Solution().combinationSum2([10,1,2,7,6,1,5], 8))

File: Problems/040-combination-sum-ii/combination-sum-ii.go
package problem040

import "sort"

func combinationSum2(candidates []int, target int) [][]int {
	var res [][]int

	sort.Ints(candidates)
	backtrack(candidates, target, 0, []int{}, &res)

	return res
}

func backtrack(candidates []int, target, start int, path []int, res *[][]int) {
	if target == 0 {
		*res = append(*res, path)
		return
	}

	for i := start; i < len(candidates); i++ {
		if i > start && candidates[i] == candidates[i-1] {
			continue
		}
		if target-candidates[i] < 0 {
			break
		}
		backtrack(candidates, target-candidates[i], i+1, append(path, candidates[i]), res)
	}
}
